core/src/main/java/com/sequenceiq/cloudbreak/core/flow2/cluster/stopstartds/StopStartDownscaleFlowService.java

```diff
@@ -0,0 +1,116 @@
+import com.sequenceiq.cloudbreak.service.stack.StackService;
+
+@Component
+public class StopStartDownscaleFlowService {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(StopStartDownscaleFlowService.class);
+
+    @Inject
+    private StackService stackService;
+
+    @Inject
+    private ClusterService clusterService;
+
+    @Inject
+    private CloudbreakFlowMessageService flowMessageService;
+
+    @Inject
+    private StackUpdater stackUpdater;
+
+    @Inject
+    private InstanceMetaDataService instanceMetaDataService;
+
+    public void clusterDownscaleStarted(long stackId, String hostGroupName, Set<Long> privateIds) {
+        clusterService.updateClusterStatusByStackId(stackId, DetailedStackStatus.DOWNSCALE_BY_STOP_IN_PROGRESS);
+        flowMessageService.fireEventAndLog(stackId, Status.UPDATE_IN_PROGRESS.name(), CLUSTER_SCALING_STOPSTART_DOWNSCALE_INIT, hostGroupName);
+
+        // TODO CB-15153: Change the message once an adjustment based downscale is supported.
+        LOGGER.debug("stopstart scaling Decommissioning from group: {}, privateIds: [{}]", hostGroupName, privateIds);
+        Stack stack = stackService.getByIdWithListsInTransaction(stackId);
+        List<String> decomissionedHostNames = stackService.getHostNamesForPrivateIds(stack.getInstanceMetaDataAsList(), privateIds);
+        LOGGER.debug("stopstart scaling Decommissioning from group: {}, hostnames: [{}]", hostGroupName, decomissionedHostNames);
+        flowMessageService.fireInstanceGroupEventAndLog(stackId, Status.UPDATE_IN_PROGRESS.name(), hostGroupName,
+                CLUSTER_SCALING_STOPSTART_DOWNSCALE_STARTING,
+                String.valueOf(decomissionedHostNames.size()), hostGroupName, String.join(",", decomissionedHostNames));
+    }
+
+    public void logCouldNotDecommission(long stackId, List<String> notDecommissionedFqdns) {
+        // TODO CB-14929: CB-15418 This needs to be an orange message (i.e. not success, not failure). Need to figure out how the UI
+        //  processes these and applies icons.
+        flowMessageService.fireEventAndLog(stackId, UPDATE_FAILED.name(), CLUSTER_SCALING_STOPSTART_DOWNSCALE_COULDNOTDECOMMISSION,
+                String.valueOf(notDecommissionedFqdns.size()), notDecommissionedFqdns.stream().collect(Collectors.joining(", ")));
+    }
+
+    public void clusterDownscalingStoppingInstances(long stackId, String hostGroupName, Set<String> decommissionedFqdns) {
+        flowMessageService.fireInstanceGroupEventAndLog(stackId, Status.UPDATE_IN_PROGRESS.name(), hostGroupName,
+                CLUSTER_SCALING_STOPSTART_DOWNSCALE_NODE_STOPPING,
+                String.valueOf(decommissionedFqdns.size()), hostGroupName, String.join(",", decommissionedFqdns));
+    }
+
+    public void instancesStopped(long stackId, List<InstanceMetaData> instancesStopped) {
+        instancesStopped.stream().forEach(x -> instanceMetaDataService.updateInstanceStatus(x, InstanceStatus.STOPPED));
```

Comment:
i mean stream() method can be removed and instancesStopped.forEach(

Reply:
Ack. I'm going to look at this, and other changes to StopStart*FlowService in [state machine error handling](https://github.com/hortonworks/cloudbreak/pull/12183) - to avoid merge conflicts in future patches.

Reply:
Fixing as part of next push. Resolving.